ci2/ci/ci.py

```diff
@@ -40,40 +48,61 @@
                     {
                         'number': pr.number,
                         'title': pr.title,
-                        'state': pr.state
+                        'batch_id': pr.batch.id if pr.batch else None,
+                        'passing': pr.passing,
+                        'review_state': pr.review_state
                     }
                     for pr in wb.prs.values()
                 ] if wb.prs else None}
             for wb in watched_branches
         ]}
 
+@routes.get('/batches/{batch_id}/log')
+@aiohttp_jinja2.template('batch_log.html')
+async def get_batch_log(request):
+    batch_id = int(request.match_info['batch_id'])
+
+    batch_client = request.app['batch_client']
+    batch = await batch_client.get_batch(batch_id)
+    status = await batch.status()
+    return {
+        'batch_id': batch_id,
+        'batch_status': status
+    }
+
 @routes.get('/healthcheck')
 async def healthcheck(request):
     return web.Response(status=200)
 
-async def refresh_loop():
-    async with aiohttp.ClientSession() as session:
-        # FIXME
-        gh = gh_aiohttp.GitHubAPI(session, 'ci2', oauth_token=oauth_token)
-        while True:
-            try:
-                for wb in watched_branches:
-                    log.info(f'refreshing {wb.branch}')
-                    await wb.refresh(gh)
-                await asyncio.sleep(60)
-            except concurrent.futures.CancelledError:
-                raise
-            except Exception as e:
-                log.error(f'{wb.branch} refresh due to exception: {traceback.format_exc()}{e}')
+async def refresh_loop(app):
+    gh = app['github_client']
+    batch = app['batch_client']
+    while True:
+        try:
+            for wb in watched_branches:
+                log.info(f'refreshing {wb.branch}')
+                await wb.refresh(gh)
+                await wb.heal(batch)
+            await asyncio.sleep(60)
+        except concurrent.futures.CancelledError:
+            raise
+        except Exception as e:
+            log.error(f'{wb.branch} refresh due to exception: {traceback.format_exc()}{e}')
 
 app.add_routes(routes)
 
 aiohttp_jinja2.setup(app,
     loader=jinja2.FileSystemLoader('ci/templates'))
 
 async def on_startup(app):
-    asyncio.ensure_future(refresh_loop())
+    asyncio.ensure_future(refresh_loop(app))
```

Comment:
ah I didn't realize you could do this. I fixed up batch to async everything except the k8s watch, but I incorporated other coroutines like this:
```python
    try:
        app.add_routes(routes)
        loop.run_until_complete(asyncio.gather(
            _run_app(app, host='0.0.0.0', port=port),
            polling_event_loop(loop)))
    finally:
        _cancel_all_tasks(loop)
        loop.run_until_complete(loop.shutdown_asyncgens())
```

I think your approach here is probably better.

Reply:
One difference is you try to clean up here (cancel tasks, shutdown_asyncgens), and I don't.  I don't know if that's important?  Maybe I should record the refresh loop and try to cancel it on shutdown?

Reply:
`app.run_app(...)` includes clean up logic for you